test/fuzz/utils.hpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (C) 2023 Intel Corporation
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+// See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <fstream>
+#include <fuzzer/FuzzedDataProvider.h>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace fuzz {
+
+enum FuzzerAPICall : uint8_t {
+    UR_PLATFORM_GET,
+    UR_DEVICE_GET,
+    UR_DEVICE_RELEASE,
+    UR_CONTEXT_CREATE,
+    UR_CONTEXT_RELEASE,
+    UR_USM_POOL_CREATE_HOST,
+    UR_USM_POOL_CREATE_DEVICE,
+    UR_USM_POOL_RELEASE_HOST,
+    UR_USM_POOL_RELEASE_DEVICE,
+    UR_USM_HOST_ALLOC_POOL,
+    UR_USM_HOST_ALLOC_NO_POOL,
+    UR_USM_DEVICE_ALLOC_POOL,
+    UR_USM_DEVICE_ALLOC_NO_POOL,
+    UR_USM_FREE_HOST_POOL,
+    UR_USM_FREE_HOST_NO_POOL,
+    UR_USM_FREE_DEVICE_POOL,
+    UR_USM_FREE_DEVICE_NO_POOL,
+    UR_PROGRAM_CREATE_WITH_IL,
+    kMaxValue = UR_PROGRAM_CREATE_WITH_IL,
+};
+
+struct Alloc {
+    ur_context_handle_t context;
+    void *ptr;
+
+    Alloc(ur_context_handle_t context, void *ptr)
+        : context(context), ptr(ptr) {}
+    ~Alloc() { urUSMFree(context, ptr); }
+};
+using Allocs = std::vector<std::unique_ptr<Alloc>>;
+
+struct Pool {
+    ur_usm_pool_handle_t handle;
+    Allocs allocs;
+
+    Pool(ur_usm_pool_handle_t handle) : handle(handle) {}
+    ~Pool() {
+        allocs.clear();
+        urUSMPoolRelease(handle);
+    }
+};
+using Pools = std::vector<std::unique_ptr<Pool>>;
+
+struct Context {
+    ur_context_handle_t handle;
+    Pools host_pools;
+    Pools device_pools;
+    Allocs no_pool_host_allocs;
+    Allocs no_pool_device_allocs;
+
+    Context(ur_context_handle_t handle) : handle(handle) {}
+    ~Context() {
+        host_pools.clear();
```

Comment:
nit: no need to call `.clear()`

Reply:
It's for calling Alloc dtors before removing Context, avoiding use-after-free.

Reply:
Ok, got it. Btw, you can also wrap `ur_context_handle_t` in unique_ptr and then you wouldn't need to implement the dtor at all. But both ways work